core/src/main/scala/pureconfig/ProductHint.scala

```diff
@@ -0,0 +1,40 @@
+package pureconfig
+
+/**
+ * A trait that can be implemented to customize how case classes are read from and written to a config.
+ *
+ * @tparam T the type of case class for which this hint applies
+ */
+trait ProductHint[T] {
+
+  /**
+   * The mapping used to map field names to config keys
+   */
+  def fieldMapping: ConfigFieldMapping
```

Comment:
Do you think it still makes sense for `ConfigFieldMapping` to exist as an entity? I think we could simplify the interface if `ProductHint` only had a `def configFieldName(typeFieldName: String): String` method. We should still provide helpers to create `ProductHint`s from two `NamingConvention`s, as is currently available for `ConfigFieldMapping`s.